pageserver/src/tenant/remote_timeline_client/download.rs

```diff
@@ -252,6 +252,104 @@ pub(super) async fn download_index_part(
     Ok(index_part)
 }
 
+pub(super) async fn download_index_part(
+    storage: &GenericRemoteStorage,
+    tenant_id: &TenantId,
+    timeline_id: &TimelineId,
+    my_generation: Generation,
+) -> Result<IndexPart, DownloadError> {
+    if my_generation.is_none() {
+        // Operating without generations: just fetch the generation-less path
+        return do_download_index_part(storage, tenant_id, timeline_id, my_generation).await;
+    }
+
+    let previous_gen = my_generation.previous();
+    let r_previous = do_download_index_part(storage, tenant_id, timeline_id, previous_gen).await;
+
+    match r_previous {
+        Ok(index_part) => {
+            tracing::debug!("Found index_part from previous generation {previous_gen:?}");
+            return Ok(index_part);
+        }
+        Err(e) => {
+            if matches!(e, DownloadError::NotFound) {
+                tracing::debug!("No index_part found from previous generation {previous_gen:?}, falling back to listing");
+            } else {
+                return Err(e);
+            }
+        }
+    };
+
+    /// Given the key of an index, parse out the generation part of the name
+    fn parse_generation(path: RemotePath) -> Option<Generation> {
+        let file_name = match path.get_path().file_name() {
+            Some(f) => f,
+            None => {
+                // Unexpected: we should be seeing index_part.json paths only
+                tracing::warn!("Malformed index key {0}", path);
+                return None;
+            }
+        };
+
+        let file_name_str = match file_name.to_str() {
+            Some(s) => s,
+            None => {
+                tracing::warn!("Malformed index key {0}", path);
+                return None;
+            }
+        };
+
+        match file_name_str.split_once('-') {
+            Some((_, gen_suffix)) => u32::from_str_radix(gen_suffix, 16)
```

Comment:
This piece should be part of `impl Generation`, living next to `fn get_suffix`

Reply:
Moved in 2649114ff125671fec40bc7ca0c3fcebf25f19b7